src/uct/tcp/sockcm/sockcm_iface.c

```diff
@@ -89,29 +106,206 @@ static uct_iface_ops_t uct_sockcm_iface_ops = {
 
 void uct_sockcm_iface_client_start_next_ep(uct_sockcm_iface_t *iface)
 {
-    ucs_trace("sockcm_iface_client_start_next_ep not implemented");
+    ucs_status_t status;
+    uct_sockcm_ep_t *ep, *tmp;
+
+    UCS_ASYNC_BLOCK(iface->super.worker->async);
+
+    /* try to start an ep from the pending eps list */
+    ucs_list_for_each_safe(ep, tmp, &iface->pending_eps_list, list_elem) {
+        status = uct_sockcm_ep_set_sock_id(iface, ep);
+        if (status != UCS_OK) {
+            continue;
+        }
+
+        ucs_list_del(&ep->list_elem);
+        ep->is_on_pending = 0;
+
+        uct_sockcm_ep_set_failed(&iface->super.super, &ep->super.super, status);
+    }
+
+    UCS_ASYNC_UNBLOCK(iface->super.worker->async);
+}
+
+static void uct_sockcm_iface_process_conn_req(uct_sockcm_iface_t *iface,
+                                             uct_sockcm_conn_param_t conn_param)
+{
+    iface->conn_request_cb(&iface->super.super, iface->conn_request_arg, NULL,
+			   conn_param.private_data, conn_param.private_data_len);
```

Comment:
Missing full context here. Will get back on this

Reply:
did you get a chance to come back to this?

Reply:
RDMACM uses the following:
```C
static void uct_rdmacm_iface_process_conn_req(uct_rdmacm_iface_t *iface,
                                              struct rdma_cm_event *event,
                                              struct sockaddr *remote_addr)
{
    uct_rdmacm_priv_data_hdr_t *hdr;

    hdr = (uct_rdmacm_priv_data_hdr_t*) event->param.ud.private_data;
    ucs_assert(hdr->status == UCS_OK);

    /* TODO check the iface's cb_flags to determine when to invoke this callback.
     * currently only UCT_CB_FLAG_ASYNC is supported so the cb is invoked from here */
    iface->conn_request_cb(&iface->super.super, iface->conn_request_arg,
                           /* connection request*/
                           event,
                           /* private data */
                           UCS_PTR_BYTE_OFFSET(event->param.ud.private_data,
                                               sizeof(uct_rdmacm_priv_data_hdr_t)),
                           /* length */
                           hdr->length);
}
```
SOCKCM uses the following instead:
```C
static void uct_sockcm_iface_process_conn_req(uct_sockcm_iface_t *iface,
                                             uct_sockcm_conn_param_t conn_param)
{
    int *sock_fd;
    sock_fd = ucs_malloc(sizeof(int), "sock_fd");

    *sock_fd = conn_param.fd;
    ucs_debug("process conn req: accepted fd %d %m", *sock_fd);

    iface->conn_request_cb(&iface->super.super, iface->conn_request_arg, sock_fd,
               conn_param.private_data, conn_param.private_data_len);
}
```
But I also set `conn_param.private_data_len` to have the same value as `hdr->length` [here](https://github.com/openucx/ucx/pull/3570/files#diff-aca878e51b7d5e6eaa75cd20a1375eadR70)

Will that work?

Reply:
But `conn_param.private_data_len` should include all the sent data length, right? including the hdr itself plus the actual data. So that the transport would pass both hdr and data to the remote side.
`conn_param.private_data_len = sizeof(*hdr) + hdr->length`
but for the callback, the upper layer only cares about the actual data: hdr->length

Reply:
Seems like `private_data_len` field is relevant only in `rdmacm` and not in ucp. Is that correct?

I had this field left as earlier on as `sockcm` is mostly derived from `rdmacm` uct. If I understand correctly, what is relevant is the length of the packed addresses that are sent over to the server to pass as the last argument to `iface->conn_request_cb` (which is represented using `hdr->length` in the case of `rdmacm` and now `sockcm` as well). If yes, then I'll plan to get rid of `conn_param.private_data_len` altogether in sockcm (because it's not used as scuh) and just keep `hdr->length` and use that in the callback instead of `conn_param.private_data_len`. Lmk if my understanding is right here and if I can proceed. Thanks!